aten/src/ATen/native/cuda/LinearAlgebra.cu

```diff
@@ -84,19 +84,26 @@ __global__ void _elementwise_kernel(int total_n_elems, func_t f) {
   }
 }
 
-template <int n_threads, int n_elems_per_thread, typename func_t>
+template <typename func_t>
 static void _launch_kernel(int total_n_elems, func_t f) {
   TORCH_INTERNAL_ASSERT(
     total_n_elems >= 0 && total_n_elems <= std::numeric_limits<int32_t>::max()
   );
 
+  const int n_threads = at::cuda::warp_size() * 2;
+  constexpr int n_elems_per_thread = 4;
   dim3 block(n_threads);
-  constexpr int total_work_block = n_threads * n_elems_per_thread;
+  const int total_work_block = n_threads * n_elems_per_thread;
   dim3 grid((total_n_elems + total_work_block - 1) / total_work_block);
 
   auto stream = at::cuda::getCurrentCUDAStream();
-  _elementwise_kernel<n_threads, n_elems_per_thread, func_t>
-    <<<grid, block, 0, stream>>>(total_n_elems, f);
+  if (n_threads == 64) {
+    _elementwise_kernel<64, n_elems_per_thread, func_t>
+        <<<grid, block, 0, stream>>>(total_n_elems, f);
+  } else if (n_threads == 128) {
+    _elementwise_kernel<128, n_elems_per_thread, func_t>
+        <<<grid, block, 0, stream>>>(total_n_elems, f);
+  }
```

Comment:
Should we guard this with ifdefs on ROCm? This increases the CUDA binary size.

Reply:
If `C10_WARP_SIZE` works correctly for device code, then perhaps template parameter could be changed to `n_warps` and the kernel can calculate `n_threads` internally? That would avoid ROCm compiling two kernels for the same GPU architecture as well.

Reply:
Elementwise kernel doesn't (or shouldn't, I didn't check) have warp-specialized code, so for both ROCm and cuda it should be compiled with 128 threads.

Reply:
see comment on scattergatherkernel.cu